internal/provider/job_resource_test.go

```diff
@@ -528,3 +531,164 @@ func testAccDeleteJob(jobUrl *string) func(s *terraform.State) error {
+		}
+	})
+
+	// Test that non-final state returns retryable error
+	t.Run("returns retryable error for non-final state", func(t *testing.T) {
+		model := &JobResourceModel{
+			URL:    types.StringValue("/api/v2/jobs/1/"), // MockConfig has "running" status
+			Status: types.StringValue("pending"),
+		}
+
+		mockClient := NewMockHTTPClient([]string{"GET"}, 200)
+		retryFunc := retryUntilAAPJobReachesAnyFinalState(mockClient, model)
+		err := retryFunc()
+
+		// Should return retryable error since "running" is not a final state
+		if err == nil {
+			t.Errorf("expected error but got none")
+		}
+		// Model should be updated with "running" status from mock response
+		if model.Status.ValueString() != "running" {
+			t.Errorf("expected status 'running', got '%s'", model.Status.ValueString())
+		}
+		// Error should indicate non-final state
+		errStr := fmt.Sprintf("%v", err)
+		if !strings.Contains(errStr, "hasn't yet reached a final state") {
+			t.Errorf("expected error to contain 'hasn't yet reached a final state', got: %v", errStr)
+		}
+	})
+
+	// Test job state transition from running to successful
+	t.Run("handles job state transition from running to successful", func(t *testing.T) {
+		model := &JobResourceModel{
+			URL:    types.StringValue("/api/v2/jobs/transition/"),
+			Status: types.StringValue("pending"),
+		}
+
+		// Track number of calls to simulate job progression
+		callCount := 0
+
+		// Create a custom mock client that changes response based on call count
+		mockClient := &MockHTTPClientWithCallCount{
+			callCount: &callCount,
+		}
+
+		retryFunc := retryUntilAAPJobReachesAnyFinalState(mockClient, model)
+
+		// First call - job should be running (returns retryable error)
+		err1 := retryFunc()
+		if err1 == nil {
+			t.Errorf("expected retryable error for running job but got none")
+		}
+		if model.Status.ValueString() != "running" {
+			t.Errorf("expected status 'running' after first call, got '%s'", model.Status.ValueString())
+		}
+
+		// Second call - job should be successful (returns no error)
+		err2 := retryFunc()
+		if err2 != nil {
+			t.Errorf("expected no error for successful job but got: %v", err2)
+		}
+		if model.Status.ValueString() != "successful" {
+			t.Errorf("expected status 'successful' after second call, got '%s'", model.Status.ValueString())
+		}
+	})
+}
+
+// Custom mock client that changes response based on call count
+type MockHTTPClientWithCallCount struct {
+	callCount *int
+}
+
+func (m *MockHTTPClientWithCallCount) Get(_ string) ([]byte, diag.Diagnostics) {
+	*m.callCount++
+
+	var response []byte
+	if *m.callCount == 1 {
+		// First call: job is running
+		response = []byte(`{"status": "running", "url": "/api/v2/jobs/transition/", "type": "run"}`)
+	} else {
+		// Subsequent calls: job is successful
+		response = []byte(`{"status": "successful", "url": "/api/v2/jobs/transition/", "type": "run"}`)
+	}
+
+	return response, diag.Diagnostics{}
```

Comment:
It's a bit surprising that the `MockHTTPClientWithCallCount` hard-codes the specific response order. It also makes the calling test a bit hard to follow, since the mocked return data is down here inside the struct.

I think if we keep it generic then we should provide the mocked return data in the constructor. But I don't know that we need to keep it generic. So I'd suggest a different name like `MockJobStatusTransitionHttpClient` so that it's clear what it's for.

I don't think we need a generic struct for this purpose

Reply:
Also, same question about the `/api/v2/jobs/transition/` URL. Is there a reason we're not using a number?